To-one relationship support

caiyue1993 opened the pull request:
### Goal
Support the to-one relationship in Realm Database. For the details of relationships, see the [doc](https://realm.io/docs/swift/latest/#many-to-one).

### Implementation
Save the to-one relationship as [CKReference](https://developer.apple.com/documentation/cloudkit/ckrecord/reference). As the [Advanced CloudKit](https://developer.apple.com/videos/play/wwdc2014/231) suggests, we use **back references** instead of forward references.

### How about to-many relationship?
Actually we could use **back references** to implement to-many relationships which is also Apple suggests. And we could fetch owned items through the following example code:
```
let predicate = NSPredicate(format: "owner.name = %@", "Jim")
let jimsDogs = realm.objects(Dog.self).filter(predicate)    
```

(And #68 also provides a way to handle to-many relationship, but is not recommended)

dbmrq commented:
This is looking great, thank you! 🎉 

Doesn't it work with Realm's [`LinkingObjects`](https://realm.io/docs/swift/latest/#relationships) though? Why use a predicate, like in your example?

caiyue1993 commented:
@dbmrq Yes, `LinkingObjects` has a dependency on the to-many relationships which not currently suppoted yet. So it can't work for now.
The predicate provides a way of using `back references` instead of `forward references` with to-many relationship. Which means, we could implement this relationship without using `List`.

dbmrq commented:
@caiyue1993 I still don't understand why this can't work with `LinkingObjects` for many-to-one relationships, though.

Here's what I'm thinking, please correct me if I'm wrong:

- A dog can have a single owner, so the `Dog` class has an `owner` property stored in Realm
- An owner can have many dogs, so it uses `LinkingObjects` to get every `Dog` whose owner is itself
- IceCream doesn't need to store an owner's dogs with CloudKit; it only needs to store each dog's owner
- When a new dog is created, it's relationship to its owner is stored in CloudKit using `CKReference`
- When Realm is synced with CloudKit, the new dog is fetched and its owner is set appropriately
- If the new dog has the correct owner, `LinkingObjects(fromType: Dog.self, property: "owner")` should get all the owner's dogs
- If the dogs can be fetched using a predicate based on the owner, the dog's owner is set correctly, so dogs can also be fetched with `LinkingObjects`
- IceCream doesn't have to deal with `LinkingObjects` directly (to create new entries). If it already sets the correct owner to a dog, `LinkingObjects` can be used to represent the inverse relationship.

Is this correct? Thanks!

caiyue1993 commented:
Hi @dbmrq , you're nothing wrong and totally right! 
I've updated my answer above and added the `LinkingObjects` to the example project. It was my misunderstanding of `LinkingObjects`(which I thought it's just working with the `List` type in Realm). Actually `LinkingObjects` works just like you summarized. 
Thanks for mentioning and correcting me 😄

dbmrq commented:
Awesome, thank you!

You can claim the bounty I left here, then: https://www.bountysource.com/issues/52087619-does-icecream-support-ckreference-realm-s-linkingobjects

🍻
